### DISCUSSION TITLE
PowerShell logging with OTLP exporter

### DISCUSSION THREAD

**Discussion by OneCyrus** [open-telemetry/opentelemetry-dotnet#4624] [Q&A]
has anyone already done some powershell logging to an OTLP endpoint?

it looks like the libraries in this repo are mainly targeted to log through the ILogger interface and depndency injection. But could there be an easy way to build on top of this library to manually send log data to an OTLP (grafana agent/loki) endpoint?

or has anyone already created a powershell module for this or has it integrated with e.g. powershell transcript?

any pointer would be great.

**Comment by davidwallis** [comment]
I was looking for something like this and struggling - so adding this for anyone that comes looking (btw this was in pwsh)

There is a start of a lib here that ShadeNofziger has done which looks promising, just lacks the logging aspect. [Repo Here](https://github.com/ShaydeNofziger/PoshOtel/)

After a lot of head scratching I finally realised I was failing because of the extension methods, so this code will allow you to create a log entry:

```powershell
using assembly Microsoft.Extensions.Logging.Abstractions.dll
using assembly Microsoft.Extensions.Logging.dll
using assembly OpenTelemetry.dll
using assembly OpenTelemetry.Exporter.Console.dll

$factory = [Microsoft.Extensions.Logging.LoggerFactory]::Create({
    param($builder)
    [Microsoft.Extensions.Logging.OpenTelemetryLoggingExtensions]::AddOpenTelemetry($builder, {
        param($opt)
        [OpenTelemetry.Logs.ConsoleExporterLoggingExtensions]::AddConsoleExporter($opt)
    })
})

$logger = $factory.CreateLogger('MyLogger')

$logger.Log([Microsoft.Extensions.Logging.LogLevel]::Information,123,"Body",$null,$null)
$logger.Log([Microsoft.Extensions.Logging.LogLevel]::Information,123,"Body","Exception",$null)
```

because of issues downloading nuget packages with powershell this will create a simple class library and then pack the required dll's into the current folder

```powershell

dotnet new classlib -o $PSScriptRoot\assemblies -n assemblies --force -f net7.0

dotnet add "$PSScriptRoot\assemblies\assemblies.csproj" package Microsoft.Extensions.Logging --version 6.0.0
dotnet add "$PSScriptRoot\assemblies\assemblies.csproj" package Microsoft.Extensions.Logging.Console --version 6.0.0
dotnet add "$PSScriptRoot\assemblies\assemblies.csproj" package OpenTelemetry --version 1.6.0
dotnet add "$PSScriptRoot\assemblies\assemblies.csproj" package OpenTelemetry.Exporter.Console --version 1.6.0

dotnet publish $PSScriptRoot\assemblies\assemblies.csproj --configuration Release --output $PSScriptRoot

```

Hope this helps as a starter - please share any further info as I see a lot of benefit in having otel data from within some of our orchestration

**Reply by georgedavid29** [reply]
while this may have been easier for folks with more coding experience, I was looking for a slightly (what i thought was more simple) option. 
I tried to post trace/span via an HTTP request to the /v1/traces endpoint of my LogicMonitor OTEL collector. 

leveraging the format included here, I created a PSCustomObject that mirrors this format: 
https://github.com/open-telemetry/opentelemetry-proto/blob/v1.4.0/examples/trace.json

I'm currently authentication errors, I'm providing the bearer token in the header, but need to see if my cloud vendor can help me figure out why I can't authenticate to it.
